hpx/runtime/threads/policies/thread_queue_mc.hpp

```diff
@@ -0,0 +1,1153 @@
+#include <hpx/assertion.hpp>
+#include <hpx/error_code.hpp>
+#include <hpx/runtime/config_entry.hpp>
+#include <hpx/runtime/threads/policies/lockfree_queue_backends.hpp>
+#include <hpx/runtime/threads/policies/queue_helpers.hpp>
+#include <hpx/runtime/threads/thread_data.hpp>
+#include <hpx/runtime/threads/policies/thread_queue.hpp>
+#include <hpx/throw_exception.hpp>
+#include <hpx/allocator_support/internal_allocator.hpp>
+#include <hpx/util/cache_aligned_data.hpp>
+#include <hpx/util/function.hpp>
+#include <hpx/util/get_and_reset_value.hpp>
+#include <hpx/timing/high_resolution_clock.hpp>
+#include <hpx/thread_support/unlock_guard.hpp>
+#include <hpx/datastructures/tuple.hpp>
+
+#ifdef HPX_HAVE_THREAD_CREATION_AND_CLEANUP_RATES
+#   include <hpx/util/tick_counter.hpp>
+#endif
+
+#include <boost/lexical_cast.hpp>
+
+#include <atomic>
+#include <cstddef>
+#include <cstdint>
+#include <exception>
+#include <functional>
+#include <list>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+///////////////////////////////////////////////////////////////////////////////
+namespace hpx { namespace threads { namespace policies
+{
+#ifdef HPX_HAVE_THREAD_QUEUE_WAITTIME
+    ///////////////////////////////////////////////////////////////////////////
+    // We control whether to collect queue wait times using this global bool.
+    // It will be set by any of the related performance counters. Once set it
+    // stays set, thus no race conditions will occur.
+    extern HPX_EXPORT bool maintain_queue_wait_times;
+#endif
+#ifdef HPX_HAVE_THREAD_MINIMAL_DEADLOCK_DETECTION
+    ///////////////////////////////////////////////////////////////////////////
+    // We globally control whether to do minimal deadlock detection using this
+    // global bool variable. It will be set once by the runtime configuration
+    // startup code
+    extern bool minimal_deadlock_detection;
+#endif
+
+    ///////////////////////////////////////////////////////////////////////////
+    // // Queue back-end interface:
+    //
+    // template <typename T>
+    // struct queue_backend
+    // {
+    //     typedef ... container_type;
+    //     typedef ... value_type;
+    //     typedef ... reference;
+    //     typedef ... const_reference;
+    //     typedef ... size_type;
+    //
+    //     queue_backend(
+    //         size_type initial_size = ...
+    //       , size_type num_thread = ...
+    //         );
+    //
+    //     bool push(const_reference val);
+    //
+    //     bool pop(reference val, bool steal = true);
+    //
+    //     bool empty();
+    // };
+    //
+    // struct queue_policy
+    // {
+    //     template <typename T>
+    //     struct apply
+    //     {
+    //         typedef ... type;
+    //     };
+    // };
+    template <typename Mutex = std::mutex,
+        typename PendingQueuing = lockfree_lifo,
+        typename StagedQueuing = lockfree_lifo,
```

Comment:
This template parameter needs to go (just like in the template for `shared_priority_queue_scheduler`)

Reply:
What I mean here is that `shared_priority_queue_scheduler` passes its `TerminatedQueuing` template parameter as `StagedQueuing` to `thread_queue_mc`.